lib/types.nix

```diff
@@ -710,6 +710,23 @@ rec {
         in mergedOption.type;
     };
 
+    # A type that is one of several submodules, similiar to types.oneOf but is usable inside attrsOf or listOf
+    # submodules need an option with a type str which is used to find the corresponding type
+    taggedSubmodules =
+      { types
+      , specialArgs ? {}
+      }: mkOptionType rec {
+      name = "taggedSubmodules";
+      description = "one of ${concatStringsSep "," (attrNames types)}";
+      check = x: if x ? type then types.${x.type}.check x else throw "No type option set in:\n${lib.generators.toPretty {} x}";
+      merge = loc: foldl'
+        (res: def: types.${def.value.type}.merge loc [
+          (lib.recursiveUpdate { value._module.args = specialArgs; } def)
+        ])
+        { };
+      nestedTypes = types;
+    };
```

Comment:
TODO
- return the options for documentation `getSubOptions`. Return an `enum` option for the type tag, or maybe it should be more recognizable.
- type merging would be nice, but we can scope that out. If not, we should probably make it behave like `enum`, as in applying the union to the type tag, and merging modules for any overlapping definitions.